Implement photon binning

JamesWrigley opened the pull request:
This was part of MID's moonshot project, it helps to get rid of noise by binning the raw data by how many photons hit each pixel.

JamesWrigley committed:
Implement photon binning

jo-moeller commented:
I learned during the beamtime, that the naming wasn't maybe that optimal. As the value that we call threshold is actually not the real threshold in the calculations, some people got confused by that.
Maybe replacing the name threshold with "ADUs/photon" or something like that?

JamesWrigley commented:
Ah I see, replaced it with `ADUs/photon` in 185ab809.

JamesWrigley committed:
Add helper trait and functions for applying pixel-wise operations

These will be useful for future functions.

JamesWrigley commented:
I just pushed a rather large refactor. The main API change is that `binPhotons()` now takes an `out` parameter to write to.

This required changing the implementation quite a bit, which unfortunately is less flexible. The previous implementation simply iterated over the underlying buffer of the `pytensor`, and this works fine as long as only one `pytensor` is being operated on. I discovered through working on a separate thing that just iterating over the buffer is unreliable because the memory layout of the `ndarray` may not match the logical layout. For example, if a 2D `ndarray` is created with a [`NPY_ARRAY_C_CONTIGUOUS`](https://numpy.org/doc/stable/reference/c-api/array.html#c.NPY_ARRAY_C_CONTIGUOUS) layout and then `np.moveaxis()` is called on it, it will no longer be `NPY_ARRAY_C_CONTIGUOUS`. A very obvious bug in hindsight, but wonderfully tricky to debug :P

So if a C++ function is operating on two `pytensor`'s then the only safe way to iterate over the buffers is to ensure that both are contiguous and their layouts match. Otherwise the element at index `i` in one array will not match the logical location of the element at `i` in the other array and the output will be garbage. IMHO this is not a good design choice because it requires the user to ensure the contiguity of their `ndarray`'s, and this could require making a copy of the `ndarray` to have a contiguous layout, which might easily be slower than calling the C++ function. Hence the refactor in 61908013 to use [array indexing](https://xtensor.readthedocs.io/en/latest/indices.html#operators-array-index).

As I mentioned, this does make the C++ function less flexible because the code depends on the rank of the `pytensor`. I added a couple of helpers in 9b6e2d5d to simplify this. These required C++17 so I also rebased the branch onto #311 (which should be merged first).

JamesWrigley commented:
> I just pushed a rather large refactor. The main API change is that `binPhotons()` now takes an `out` parameter to write to.
> 
> This required changing the implementation quite a bit, which unfortunately is less flexible. The previous implementation simply iterated over the underlying buffer of the `pytensor`, and this works fine as long as only one `pytensor` is being operated on. I discovered through working on a separate thing that just iterating over the buffer is unreliable because the memory layout of the `ndarray` may not match the logical layout. For example, if a 2D `ndarray` is created with a [`NPY_ARRAY_C_CONTIGUOUS`](https://numpy.org/doc/stable/reference/c-api/array.html#c.NPY_ARRAY_C_CONTIGUOUS) layout and then `np.moveaxis()` is called on it, it will no longer be `NPY_ARRAY_C_CONTIGUOUS`. A very obvious bug in hindsight, but wonderfully tricky to debug :P
> 
> So if a C++ function is operating on two `pytensor`'s then the only safe way to iterate over the buffers is to ensure that both are contiguous and their layouts match. Otherwise the element at index `i` in one array will not match the logical location of the element at `i` in the other array and the output will be garbage. IMHO this is not a good design choice because it requires the user to ensure the contiguity of their `ndarray`'s, and this could require making a copy of the `ndarray` to have a contiguous layout, which might easily be slower than calling the C++ function. Hence the refactor in [6190801](https://github.com/European-XFEL/EXtra-foam/commit/61908013caf806c10770449d6869638b7c785d16) to use [array indexing](https://xtensor.readthedocs.io/en/latest/indices.html#operators-array-index).
> 
> As I mentioned, this does make the C++ function less flexible because the code depends on the rank of the `pytensor`. I added a couple of helpers in [9b6e2d5](https://github.com/European-XFEL/EXtra-foam/commit/9b6e2d5d2954800bd07c83c16f8e6c77e6402a59) to simplify this. These required C++17 so I also rebased the branch onto #311 (which should be merged first).

I've been re-thinking this, and I'm no longer sure it's the best way forward. The reason I refactored it was because I wanted to use the function in a special suite, with data that had been assembled with [`stack_detector_data`](https://extra-data.readthedocs.io/en/latest/agipd_lpd_data.html#extra_data.stack_detector_data) (which rearranges axes at some point and thus makes the array non-contiguous). However, in hindsight I think that whole thing could be done much more easily a different way, which removes the original need for this refactor.

Additionally, the C++ code here is... complicated..., and if we continue supporting non-contiguous layouts like this then all future functions we implement will also be as complicated (in addition to the existing C++ code). So, I now lean towards making a hard requirement that the input arrays have contiguous layouts, and throwing an error if they don't. I think this will make the code simpler, and also slightly faster in some cases. What do y'all think?

CammilleCC commented on extra_foam/algorithms/tests/test_image_proc.py:
```diff
@@ -5,11 +5,32 @@
+
+        result = bin_photons(arr3d, 2, out=arr3d)
+        self.assertIs(arr3d, result)
+        self.assertFalse(np.array_equal(arr3d, arr3d_copy))
```
Thanks for adding tests! :D

While there is a C++ test for the actual binning logic, it would be also great to test the Python output here for completeness.

JamesWrigley replied:
Added in 34f83680. Good thing you mentioned it, I found a wee bug related to NaN handling while adding the test :)

CammilleCC commented on extra_foam/algorithms/imageproc_py.py:
```diff
@@ -7,13 +7,28 @@
+    if out is None:
+        out = np.empty_like(data)
+
+    binPhotons(data, out, adu_count)
```
Nitpick: I would suggest following the signature as observed from the other functions: `f(arr, *args, out)`. Also, using the same array when `out` is not specified.

JamesWrigley replied:
Changed the signature in 34f83680. I didn't default to modifying the input array in-place though, that doesn't match the behaviour of numpy/scipy so I think it'd be a little confusing to do it here.

CammilleCC commented on src/extra_foam/include/f_utilities.hpp:
```diff
@@ -54,6 +60,49 @@ inline void checkEven(T a, std::string&& header)
+          }
+        }
+      }
+#if defined(FOAM_USE_TBB)
```
Question: why do we ask `defined(FOAM_USE_TBB)` here again?

JamesWrigley replied:
It's to add the closing brace and parenthesis to match the call to `tbb::parallel_for()`. The single-threaded implementation is somewhat awkwardly nested inside the parallel implementation, so that's why this needs to be checked again at the end.

CammilleCC commented on extra_foam/algorithms/imageproc_py.py:
```diff
@@ -7,13 +7,28 @@
 
+def bin_photons(data, adu_count, out=None):
+    rank = len(data.shape)
+    if rank != 2 and rank != 3:
```
Nitpick: you could use `data.ndim` here, and `rank not in (2, 3)` prolly is more descriptive.

JamesWrigley replied:
Fixed in 34f83680.

CammilleCC reviewed (commented):
Thanks for your work, I've tested it and it works great! :D 
I only have a few questions, but in general LGTM

Mmm, I would incline supporting non-contiguous arrays. I can't say much about the C++ utils that you've added, but seems that it works seamlessly. Could you also add a test for both non- and contiguous arrays just in case we go forward differently in the future?

JamesWrigley commented:
Oh wait, forgot to add a test for [non-]contiguous arrays.

JamesWrigley commented:
Alrighty, added in c4b59c37.

CammilleCC reviewed (approved):
LGTM :D
